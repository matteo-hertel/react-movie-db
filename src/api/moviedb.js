import { addLanguage, addAdultFilter, addPagination, addQuery, addApiKey } from "./transformers/url";


/**
 * High level function that will return the available function
 * to query MovieDB API
 * 
 * Because of the time constraint I might skip some tests or go
 * with questionable choices :)
 * 
 * @export
 * @param {any} config injected configs 
 * @returns Object
 */
export function MovieDb(config) {
    return {
        searchMovies: searchMovies(config), //the search movie is exportable but I'll inject the configs so this module is ready to use
        getMovie: function () { }
    };
}
/**
 * @export
 * @param {any} config requires baseUrl and apiKey
 * this will make this function working with different configs and be highly reusable
 * @returns function
 */
export function searchMovies(config) {
    /**
     * in here maybe a check to see if the required configs are there
     * it's a super small app so for now will skip that just add a comment in the
     * docs above
     */

    /**
     * Name your anonymous functions, it will be extremely helpful while debugging
     * 
     * search function will accept a query in a return a promise (using fetch) to resolve the 
     * HTTP call
     * 
     * @param string query
     * @return Promise
     */
    return function search(query) {
        /**
         * having this hardcoded URL is not good, I'll spend a bit more time adding a better way 
         * to handle the query parameters in a better way
         */
        let url = addLanguage(
            addAdultFilter(
                addPagination(
                    addQuery(
                        addApiKey(`${config.baseUrl}/search/movie`, config.apiKey, true),
                        query)
                )
            )
        )
        return fetch(url)
    }
}