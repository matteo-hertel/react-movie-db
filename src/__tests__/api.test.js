global.XMLHttpRequest = require('w3c-xmlhttprequest').XMLHttpRequest;

import { MovieDb } from "./../api/moviedb";
import config from "./../config";

const MovieAPI = MovieDb(config);
const query = encodeURIComponent("Wonder Woman");
/**
 * Test the high level module in order to have test something ready for production
 * if I have spare time I'll be able to test every single bit
 */
describe("The high level module", function () {

    it('should return an object', () => {
        expect(typeof MovieAPI).toEqual("object"); //I'm sure there is a better way to do this but for now will do
    });

    it('should have a searchMovie function that returns a promise', () => {
        // need to add a mock function for fetch, it's doing real HTTP requests ATM
        expect(typeof MovieAPI.searchMovies(query).then).toEqual("function"); //I'm sure there is a better way to do this but for now will do
        expect(typeof MovieAPI.searchMovies(query).catch).toEqual("function"); //I'm sure there is a better way to do this but for now will do
    });
})