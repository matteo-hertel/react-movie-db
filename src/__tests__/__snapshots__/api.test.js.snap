// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The high level module should have a getMovie function 1`] = `
Object {
  "adult": false,
  "backdrop_path": "/hA5oCgvgCxj5MEWcLpjXXTwEANF.jpg",
  "belongs_to_collection": null,
  "budget": 120000000,
  "genres": Array [
    Object {
      "id": 28,
      "name": "Action",
    },
    Object {
      "id": 12,
      "name": "Adventure",
    },
    Object {
      "id": 14,
      "name": "Fantasy",
    },
    Object {
      "id": 878,
      "name": "Science Fiction",
    },
  ],
  "homepage": "http://www.warnerbros.com/wonder-woman",
  "id": 297762,
  "imdb_id": "tt0451279",
  "original_language": "en",
  "original_title": "Wonder Woman",
  "overview": "An Amazon princess comes to the world of Man to become the greatest of the female superheroes.",
  "popularity": 240.134252,
  "poster_path": "/imekS7f1OuHyUP2LAiTEM0zBzUz.jpg",
  "production_companies": Array [
    Object {
      "id": 444,
      "name": "Dune Entertainment",
    },
    Object {
      "id": 507,
      "name": "Atlas Entertainment",
    },
    Object {
      "id": 6194,
      "name": "Warner Bros.",
    },
    Object {
      "id": 9993,
      "name": "DC Entertainment",
    },
    Object {
      "id": 9995,
      "name": "Cruel & Unusual Films",
    },
    Object {
      "id": 81620,
      "name": "TENCENT PICTURES",
    },
    Object {
      "id": 83838,
      "name": "Wanda Pictures",
    },
  ],
  "production_countries": Array [
    Object {
      "iso_3166_1": "US",
      "name": "United States of America",
    },
  ],
  "release_date": "2017-05-30",
  "revenue": 573495580,
  "runtime": 141,
  "spoken_languages": Array [
    Object {
      "iso_639_1": "de",
      "name": "Deutsch",
    },
    Object {
      "iso_639_1": "en",
      "name": "English",
    },
  ],
  "status": "Released",
  "tagline": "Power. Grace. Wisdom. Wonder.",
  "title": "Wonder Woman",
  "video": false,
  "vote_average": 7,
  "vote_count": 2302,
}
`;

exports[`The high level module should have a searchMovie function 1`] = `
Object {
  "page": 1,
  "results": Array [
    Object {
      "adult": false,
      "backdrop_path": "/hA5oCgvgCxj5MEWcLpjXXTwEANF.jpg",
      "genre_ids": Array [
        28,
        12,
        14,
        878,
      ],
      "id": 297762,
      "original_language": "en",
      "original_title": "Wonder Woman",
      "overview": "An Amazon princess comes to the world of Man to become the greatest of the female superheroes.",
      "popularity": 241.134252,
      "poster_path": "/imekS7f1OuHyUP2LAiTEM0zBzUz.jpg",
      "release_date": "2017-05-30",
      "title": "Wonder Woman",
      "video": false,
      "vote_average": 7,
      "vote_count": 2293,
    },
    Object {
      "adult": false,
      "backdrop_path": "/uC5s9lCzpvHzJ5YnF26cjOfDpsZ.jpg",
      "genre_ids": Array [
        28,
        12,
        16,
        878,
      ],
      "id": 15359,
      "original_language": "en",
      "original_title": "Wonder Woman",
      "overview": "On the mystical island of Themyscira, a proud and fierce warrior race of Amazons have raised a daughter of untold beauty, grace and strength: Princess Diana. When an Army fighter pilot, Steve Trevor, crash-lands on the island, the rebellious and headstrong Diana defies Amazonian law by accompanying Trevor back to civilization.",
      "popularity": 2.498122,
      "poster_path": "/u6sI0yLJXkmrbSJzVRCmmfa4nhx.jpg",
      "release_date": "2009-03-03",
      "title": "Wonder Woman",
      "video": false,
      "vote_average": 7.1,
      "vote_count": 112,
    },
    Object {
      "adult": false,
      "backdrop_path": null,
      "genre_ids": Array [
        28,
        14,
      ],
      "id": 241224,
      "original_language": "en",
      "original_title": "Wonder Woman",
      "overview": "It's one hundred and fifty seconds of pure fantasy.",
      "popularity": 1.325837,
      "poster_path": "/6xrVlKEGBTJCPhTuBfADvknrh5D.jpg",
      "release_date": "2013-09-29",
      "title": "Wonder Woman",
      "video": false,
      "vote_average": 6.2,
      "vote_count": 12,
    },
    Object {
      "adult": false,
      "backdrop_path": "/kq6qLxKGUFHZUJBdF5Y4HfUr692.jpg",
      "genre_ids": Array [
        28,
        12,
        14,
        10770,
      ],
      "id": 161620,
      "original_language": "en",
      "original_title": "Wonder Woman",
      "overview": "A super-hero uses her powers to thwart an international spy ring.",
      "popularity": 1.203247,
      "poster_path": "/NwG3qksGquPDTEwILUm8EfQ6BA.jpg",
      "release_date": "1974-03-12",
      "title": "Wonder Woman",
      "video": false,
      "vote_average": 5.3,
      "vote_count": 6,
    },
    Object {
      "adult": false,
      "backdrop_path": null,
      "genre_ids": Array [],
      "id": 461973,
      "original_language": "en",
      "original_title": "wonder woman",
      "overview": "Action Â· Before she was Wonder Woman she was Diana, princess of the Amazons, trained warrior. When a pilot crashes and tells of conflict in the outside world",
      "popularity": 1.001157,
      "poster_path": null,
      "release_date": "",
      "title": "wonder woman",
      "video": false,
      "vote_average": 0,
      "vote_count": 0,
    },
    Object {
      "adult": false,
      "backdrop_path": null,
      "genre_ids": Array [
        28,
      ],
      "id": 306824,
      "original_language": "en",
      "original_title": "Technology/Transformation: Wonder Woman",
      "overview": "The video opens with a barrage of explosive imagery along with an audio track of a siren taken from the 1970s TV show Wonder Woman. The following scenes are fast paced repeated shots from Wonder Woman, with several scenes following of actress Lynda Carter as the main character Diana Prince, performing her transformative spin from secretarial role into superhero role. [â¦] The representation of repeated transformations expose the illusion of fixed female identities in media and attempts to show the emergence of a new woman through use of technology. [â¦] The video ends with a scene of repeating explosions that precedes a blue background with white text that scrolls upwards, delivering a transcription of lyrics to the song âWonder Woman Disco' (1978) by The Wonderland Disco Band. (Wikipedia)",
      "popularity": 1.000001,
      "poster_path": "/zi5BFsd4pfdI2Rpx57ho9yaXMeX.jpg",
      "release_date": "1978-11-29",
      "title": "Technology/Transformation: Wonder Woman",
      "video": false,
      "vote_average": 5.8,
      "vote_count": 4,
    },
    Object {
      "adult": false,
      "backdrop_path": null,
      "genre_ids": Array [
        14,
        28,
      ],
      "id": 464052,
      "original_language": "en",
      "original_title": "Wonder Woman 2",
      "overview": "Plot unknown.",
      "popularity": 1.225,
      "poster_path": null,
      "release_date": "",
      "title": "Wonder Woman 2",
      "video": false,
      "vote_average": 0,
      "vote_count": 0,
    },
    Object {
      "adult": false,
      "backdrop_path": null,
      "genre_ids": Array [
        28,
      ],
      "id": 346578,
      "original_language": "en",
      "original_title": "Wonder Woman: Molded from Clay",
      "overview": "Wonder Woman: Molded from Clay is an action science fiction movie directed by Jason Hough",
      "popularity": 1.213293,
      "poster_path": null,
      "release_date": "2016-01-28",
      "title": "Wonder Woman: Molded from Clay",
      "video": false,
      "vote_average": 10,
      "vote_count": 1,
    },
    Object {
      "adult": false,
      "backdrop_path": null,
      "genre_ids": Array [],
      "id": 30041,
      "original_language": "en",
      "original_title": "Wonder Womyn Vs. Witchy Woman: Curse of the Kijoka Talisman",
      "overview": "No overview found.",
      "popularity": 1.000445,
      "poster_path": null,
      "release_date": "2008-01-01",
      "title": "Wonder Womyn Vs. Witchy Woman: Curse of the Kijoka Talisman",
      "video": false,
      "vote_average": 0,
      "vote_count": 0,
    },
    Object {
      "adult": false,
      "backdrop_path": null,
      "genre_ids": Array [
        35,
      ],
      "id": 295898,
      "original_language": "en",
      "original_title": "La donna Ã¨ una cosa meravigliosa",
      "overview": "Italian episode movie",
      "popularity": 1.002127,
      "poster_path": "/6aqTIAMMN45sYRWv8ChTvPGzdrP.jpg",
      "release_date": "1964-12-31",
      "title": "Woman Is a Wonderful Thing",
      "video": false,
      "vote_average": 0,
      "vote_count": 0,
    },
    Object {
      "adult": false,
      "backdrop_path": null,
      "genre_ids": Array [],
      "id": 250036,
      "original_language": "it",
      "original_title": "La donna delle meraviglie",
      "overview": "In a straightforward yet unexceptional rendering of the life and loves of Alberto (Ben Gazzara), a writer/director approaching the hill though not yet over it, director Alberto Bevilacqua has used his own experiences to fashion a conventional tale about a search for the \\"perfect woman.\\" His mettlesome neighbor Luisa (Lina Sastri) just will not leave him alone; she was the reason why he ended his marriage to Maura (Claudia Cardinale), but Alberto considers their affair dead and buried. Recently, an anonymous \\"dream woman\\" has been calling him for some seductive phone exchanges. As he searches out the identity of the mystery caller he is also distracted by Luisa who starts to insinuate herself back into his affections. Sooner or later, Alberto will just have to decide how to handle his love life, which in the end may only be mesmerizing to himself alone.",
      "popularity": 1.148872,
      "poster_path": "/7TUIIwD64XlYX7C0SuUtzJcWkfS.jpg",
      "release_date": "1985-01-13",
      "title": "Woman of Wonders",
      "video": false,
      "vote_average": 0,
      "vote_count": 0,
    },
  ],
  "total_pages": 1,
  "total_results": 11,
}
`;
